core/src/main/scala/org/apache/spark/scheduler/BlacklistStrategy.scala

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.scheduler
+
+import scala.collection.mutable
+
+import org.apache.spark.Logging
+import org.apache.spark.SparkConf
+import org.apache.spark.util.SystemClock
+
+/**
+ * The interface to determine executor blacklist and node blacklist.
+ */
+trait BlacklistStrategy {
+  val expireTimeInMillisecond: Long
+
+  // TaskId is optional here because the BlacklistStrategy could be unrelated to TaskId
+  def getExecutorBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus],
+      taskId: Option[Long]): Set[String]
+
+  def getNodeBlacklist(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus],
+      taskId: Option[Long]): Set[String]
+
+  // Default implementation to remove failure executors from HashMap based on given time period.
+  def expireExecutorsInBlackList(
+      failedExecutorMap: mutable.HashMap[String, FailureStatus]): Unit = {
+    val now = new SystemClock().getTimeMillis()
+    failedExecutorMap.retain((executorid, failureStatus) => {
+      (now - failureStatus.updatedTime) < expireTimeInMillisecond
+    })
```

Comment:
For (b),  Yes, there is logical flaw, and I agree with your suggestion. Thanks for point it out.